### ISSUE TITLE
question [python-package] pandas and categorical support

### ISSUE THREAD

**Issue by geoHeil** [Microsoft/LightGBM#192]
How can I create a `lgb.Dataset` directly from `pandas`?

`train_data = lgb.Dataset(data, label=label, feature_name=['c1', 'c2', 'c3'], categorical_feature=['c3'])
`shows how to use categorical features.
Can these be inferred directly from pandas `categorical` datatypes or must these be specified manually?

**Comment by wxchan** [comment]
cannot infer yet. it's a good feature to be added.

**Comment by guolinke** [comment]
@wxchan I think you can help to add following features in python package.
 
1. auto infer categorical features
2. convert them to int type (for all categorical features, not just inferred)

**Comment by geoHeil** [comment]
If you convert to int how would new categorical levels be handled?
Guolin Ke <notifications@github.com> schrieb am Mi. 11. Jan. 2017 um 10:39:

**Comment by guolinke** [comment]
@geoHeil After this fix, for the non-numerical type features, it will be treated as categorical features. 
For other categorical features that is int type, you still need to specific them in the ```categorical_feature```

**Comment by vruusmann** [comment]
Would it be possible to "export" categorical feature information also in `GBDT::SaveModelToFile` method? The section of feature importances could be followed (or preceded) by another section, which makes the assumption about feature type and the valid value space explicit.

For example, feature specification for the [Auto-MPG dataset](https://archive.ics.uci.edu/ml/datasets/Auto+MPG):
```
Features=7 // Header
cylinders=3 4 5 6 8 // Categorical
displacement=[68, 455] // Continuous
horsepower=[46, 230] // Continuous
weight=[1613, 5140] // Continuous
acceleration=[8.0, 24.8] // Continuous
model_year=70 71 72 73 74 75 76 77 78 79 80 81 82 // Categorical
origin=1 2 3 // Categorical
```

Perhaps, if the text file contained such feature specification section, then it would be possible to drop the `feature_name` entry from the model header section.

Please note that you can apply "ad hoc" one-hot-encoding to categorical variables. For example, the feature `cylinders` would be re-mapped from `3 4 5 6 8` value space to `1 2 3 4 5` value space (in the latter case, the `0` value should be reserved for missing values?), so that the check for a five-cylinder car would be represented as `cylinders == 3`.

**Comment by vruusmann** [comment]
The proposed "ad hoc" one-hot-encoding would also address the situation, where Pandas data column contains string values.

**Comment by geoHeil** [comment]
@vruusmann still it is unclear to me if this would only work with learned labels e.g. the ones present during training or if new ones would be supported as well. In case of the ad hoc one-hot-encoding this would mean that new (and thus unknown) columns would be generated.

**Comment by vruusmann** [comment]
Here's an example LightGBM text file for the abovementioned Auto-MPG dataset: https://github.com/jpmml/jpmml-lightgbm/blob/master/src/test/resources/lgbm/RegressionAuto.txt

It's possible to extract the set of "active" category values for some feature by observing which values are used as thresholds (ie. filter threshold values where `split_feature[i] == $feature && decision_type[i] == 1`). For example, it is easy to see that LightGBM algorithm thinks that only `4` and `6` are significant cylinder counts.

However, what is missing from the current text file is the set of "inactive" feature values. For example, it would be essential to communicate that the training dataset also contained `cylinder` values of `3`, `5` and `8`, but they weren't used as split thresholds.

The sum of "active" and "inactive" feature values defines the total valid value space for some feature. This way, it will be allowed to use this LightGBM model for prediction when the value of `cylinders` feature is one of `3`, `4`, `5`, `6` and `8`. However, if it's something else, then this LightGBM model shouldn't be used, because the incoming data record represents "uncharted territory".

Furthermore, the feature specification section would make it explicit the results of LightGBM's internal "continuous vs. categorical" inference algorithm.

**Comment by guolinke** [comment]
@vruusmann 
You want a section that contain more information about features. e.g. feature_name, feature_type and value_range, right? 
And if the feature values in prediction data is not in the valid range/value, throw errors? 

However, I think it cannot see all values of features in training data. 
In my opinion, I think it should be able to predict the data with un-seen value as well.

**Comment by vruusmann** [comment]
> You want a section that contain more information about features. e.g. feature_name, feature_type and value_range, right?

It would be very nice if LightGBM exported all information that it has about features, as that would make it easier to 3rd party applications to get a quick overview of model's "schema".

Currently, the header section only contains the list of feature names. The operational type (continuous vs categorical) can be inferred by parsing tree section(s) and checking the value of corresponding `decision_type` attributes.

> And if the feature values in prediction data is not in the valid range/value, throw errors?

There's no need to modify the behaviour of LightGBM prediction algorithm in any way. The idea is that a 3rd party application should parse the feature information section, and decide independently if the input dataset is compatible with the training dataset or not (eg. ask for user's confirmation - "your dataset contains twelve-cylinder cars, which were not present in the training dataset - are you sure you want to continue with scoring?").

**Comment by guolinke** [comment]
@vruusmann you can try the latest commit, already output more feature information in the last section of model file.
